libraries/chain/combined_database.cpp

```diff
@@ -535,57 +541,280 @@ namespace eosio { namespace chain {
+            using utils_t     = decltype(utils);
+            using value_t     = typename utils_t::index_t::value_type;
+            using by_table_id = object_to_table_id_tag_t<value_t>;
+
+            auto tid_key      = boost::make_tuple(table_row.id);
+            auto next_tid_key = boost::make_tuple(table_id_object::id_type(table_row.id._id + 1));
+
+            unsigned_int size = utils_t::template size_range<by_table_id>(db, tid_key, next_tid_key);
+            section.add_row(size, db);
+
+            utils_t::template walk_range<by_table_id>(db, tid_key, next_tid_key, [&db, &section](const auto& row) {
+               static int count = 0;
+               if (row.t_id == 7 && row.primary_key == 1)
+                  ++count;
+               section.add_row(row, db);
+            });
+         });
+      });
+   }
+
+   struct table_id_object_view {
+      account_name   code;  //< code should not be changed within a chainbase modifier lambda
+      scope_name     scope; //< scope should not be changed within a chainbase modifier lambda
+      table_name     table; //< table should not be changed within a chainbase modifier lambda
+      account_name   payer;
+      uint32_t       count = 0; /// the number of elements in the table
+   };
+
+   struct blob {
+      std::string data;
+   };
+
+   template<typename DataStream>
+   inline DataStream& operator << ( DataStream& ds, const blob& b ) {
+      fc::raw::pack(ds, b.data);
+      return ds;
+   }
+
+   template<typename DataStream>
+   inline DataStream& operator >> ( DataStream& ds, blob& b ) {
+      fc::raw::unpack(ds, b.data);
+      return ds;
+   }
+
+   struct primary_index_view {
+      uint64_t     primary_key;
+      account_name payer;
+      blob     value;
+   };
+
+   template <typename SecondaryKeyType>
+   struct secondary_index_view {
+      using secondary_key_type = SecondaryKeyType;
+      uint64_t primary_key;
+      account_name payer;
+      SecondaryKeyType secondary_key;
+   };
+
+   template <typename Section>
+   struct rocksdb_contract_table_writer {
+      using key_type = eosio::chain::backing_store::db_key_value_format::key_type;
+
+      Section&                                   section_;
+      const chainbase::database&                 db_;
+      eosio::session::undo_stack<rocks_db_type>& kv_undo_stack_;
+      std::vector<primary_index_view>            primary_indices;
+      std::tuple<std::vector<secondary_index_view<uint64_t>>,
+                 std::vector<secondary_index_view<uint128_t>>,
+                 std::vector<secondary_index_view<key256_t>>,
+                 std::vector<secondary_index_view<float64_t>>,
+                 std::vector<secondary_index_view<float128_t>> >
+          all_secondary_indices;
+
+      using extract_index_member_fun_t = void (rocksdb_contract_table_writer::*)(b1::chain_kv::bytes::const_iterator, b1::chain_kv::bytes::const_iterator, const char*, std::size_t);
```

Comment:
We should try to avoid the use of any of the chain_kv types.  Those are being deprecated and replaced by the session api.

Reply:
I'll let @b1bart determine if we have to do that now.  I'd prefer to get this all merged into develop, then work on removing those.